runtime/parachains/src/paras.rs

```diff
@@ -902,55 +1435,163 @@ impl<T: Config> Pallet<T> {
-				let next_possible_upgrade_at =
-					relay_parent_number + cfg.validation_upgrade_frequency;
+		let mut weight = T::DbWeight::get().reads(1);
 
-				*up = Some(expected_at);
+		// Enacting this should be prevented by the `can_schedule_upgrade`
+		if FutureCodeHash::<T>::contains_key(&id) {
+			// This branch should never be reached. Signalling an upgrade is disallowed for a para
+			// that already has one upgrade scheduled.
+			//
+			// Any candidate that attempts to do that should be rejected by
+			// `can_upgrade_validation_code`.
+			//
+			// NOTE: we cannot set `UpgradeGoAheadSignal` signal here since this will be reset by
+			//       the following call `note_new_head`
+			log::warn!(
+				target: "runtime::paras",
+				"ended up scheduling an upgrade while one is pending",
+			);
+			return weight
+		}
 
-				<Self as Store>::UpcomingUpgrades::mutate(|upcoming_upgrades| {
-					let insert_idx = upcoming_upgrades
-						.binary_search_by_key(&expected_at, |&(_, b)| b)
-						.unwrap_or_else(|idx| idx);
-					upcoming_upgrades.insert(insert_idx, (id, expected_at));
-				});
+		let code_hash = new_code.hash();
+
+		// para signals an update to the same code? This does not make a lot of sense, so abort the
+		// process right away.
+		weight += T::DbWeight::get().reads(1);
+		if CurrentCodeHash::<T>::get(&id) == Some(code_hash) {
+			// NOTE: we cannot set `UpgradeGoAheadSignal` signal here since this will be reset by
+			//       the following call `note_new_head`
+			log::warn!(
+				target: "runtime::paras",
+				"para tried to upgrade to the same code. Abort the upgrade",
+			);
+			return weight
+		}
 
-				// From the moment of signalling of the upgrade until the cooldown expires, the
-				// parachain is disallowed to make further upgrades. Therefore set the upgrade
-				// permission signal to disallowed and activate the cooldown timer.
-				<Self as Store>::UpgradeRestrictionSignal::insert(&id, UpgradeRestriction::Present);
-				<Self as Store>::UpgradeCooldowns::mutate(|upgrade_cooldowns| {
-					let insert_idx = upgrade_cooldowns
-						.binary_search_by_key(&next_possible_upgrade_at, |&(_, b)| b)
-						.unwrap_or_else(|idx| idx);
-					upgrade_cooldowns.insert(insert_idx, (id, next_possible_upgrade_at));
-				});
+		// This is the start of the upgrade process. Prevent any further attempts at upgrading.
+		weight += T::DbWeight::get().writes(2);
+		FutureCodeHash::<T>::insert(&id, &code_hash);
+		UpgradeRestrictionSignal::<T>::insert(&id, UpgradeRestriction::Present);
+
+		weight += T::DbWeight::get().reads_writes(1, 1);
+		let next_possible_upgrade_at = relay_parent_number + cfg.validation_upgrade_frequency;
+		<Self as Store>::UpgradeCooldowns::mutate(|upgrade_cooldowns| {
+			let insert_idx = upgrade_cooldowns
+				.binary_search_by_key(&next_possible_upgrade_at, |&(_, b)| b)
+				.unwrap_or_else(|idx| idx);
+			upgrade_cooldowns.insert(insert_idx, (id, next_possible_upgrade_at));
+		});
```

Comment:
Why not set this when the upgrade was accepted? Otherwise, if the values are choosen "unlucky", it could happen that we finish an upgrade and the next one is directly doable?

Reply:
1. In this PR, I mention [this](https://github.com/paritytech/polkadot/pull/4457/files#diff-9d731023bd72eb1b3b196e18a4c373d586ab6379f01a2029799cafed907fc8d5R78) for the documentation. We should enforce the invariant on this in #4248 or similar.
2. It is not possible as of currently, since if the upgrade was not consumed by the para (i.e. it did not produce any new heads, see `note_new_head`) then `can_upgrade_validation_code` will return `false`. This means that such a candidate will be rejected by backing and by the on-chain logic in inclusion module.
3. If we do want to address this, I think it should go in a follow-up.

Reply:
So I ended up not fixing it, see #4603.